# Dota 2
## Abstract
When determining fault for a loss, placing blame on only the worst player on a team means that it's plausible that a Dota 2 player is only to blame for 2 of their last 18 losses. Placing blame on anyone who played below their average means that it's highly unlikely that a Dota 2 player is only to blame for 2 of their last 18 losses.

## Background and Conventions
This is an examination of a discussion that I came across online. It devolved past the point of productive conversation, but I think the two involved (the original poster and the commenter) would have been able to benefit from one another's perspective if they had understood that each was working off of different definitions and beliefs.

Capitalized nouns (other than OP) refer to the character a player uses to interact with the game, e.g., Mirana, Dazzle, Tiny. These characters are also called heroes. These heroes buy items, destroy enemy structures, and march up into the enemy's stronghold to end the conflict. It requires an intense amount of cooperation between the player and their four teammates.

## A Question of Fault
On the Dota 2 subreddit, there are an enormous number of different perspectives about how to play and how to improve. Redditors, looking for validation, will often describe a game they just played, tell us all about how their team played poorly, and conclude saying, "There was nothing I could have done." Other players then respond, asking them what they were doing on their hero. "Did you buy the right items? Did you play around your team correctly? Did you pressure other lanes as much as you needed to? Did you TP when your mid was getting ganked?" The original commenter then feels attacked, and says, "It's not my fault that I lost the game. My (Sniper | Troll | Dazzle | AM | Tiny) played sooooo bad." The comments then inevitably devolve into puerile insults.

In one recent example, a commenter described how they had lost 16 of the last 18 games that they had played. He mentioned that in one of his games, a Mirana (who was supposed to be supporting him) followed him around, sniping every creep for twenty minutes. In another game, his carry threw (caused his team to lose a winning game) by failing at going high ground multiple times. Over the course of those 18 games, he said it had only been his fault twice that he'd lost.

"You've lost 16 out of your last 18 games, and there were only two games where you could have improved? You just need to admit that you're bad. That's so unlikely," a Reditor responded. The exchange then got significantly more vitriolic. As it happens, I don't think that either of these two are wrong - they're just looking at the game from two different perspectives.

## Examining OP's perspective
From the original poster's (OP's) point of view, it's his fault that he lost when he's the worst player on his team. What's the chance that he's the worst player on his team?

In [2]:
worst_player_count = 1
players_on_team = 5
probability_op_is_worst_player = 1 / 5 #Naive Bayes prior
print(probability_op_is_worst_player)

0.2


Under this point of view, the probability is 0.2, or 20%, that OP is the worst player on his team. Under that definition, what are the odds that out of 18 games, he was the worst player in two of them?

In [3]:
import math
total_games = 18

# Games where the OP was the worst player on his team
games_where_worst = 2 

# Games where the OP was NOT the worst player on his team
games_where_not = total_games - games_where_worst 

# The probability of the OP being the worst player in <games_where_worst> games in a row
worst_player_chance = probability_op_is_worst_player ** games_where_worst 

 # The probability of OP not being the worst player <games_where_not> games in a row
not_worst_player_chance = (1 - probability_op_is_worst_player) ** games_where_not

# We have the probabilities of each happening multiple times in a row,
# but we want to know the probability when they're combined.
# Since we don't have any evidence that the chance that he's the worst player
# depends on whether he was the worst player in his previous game, the chance
# that he's the worst player in x games in a row can be treated as independent
# of the chance that he's not the worst player y games in a row
# P(AB) = P(A) * P(B)  [when event A and event B are independent]
joint_chance = worst_player_chance * not_worst_player_chance

# Now, we want the probabilities when they don't happen in a particular order,
# so we need the number of different ways to play <total_games> games
ways_to_play_total_games = math.factorial(total_games)

# We don't care about the order of the games where the OP was the worst, so we
# divide by the number of ways to play <games_where_worst> games
ways_total_unordered_worst = ways_to_play_total_games / math.factorial(games_where_worst)

# We also don't care about the order of the games where OP wasn't the worst, so
# we divide by the number of ways to play <games_where_not> games
ways_total_unordered_both = ways_total_unordered_worst/math.factorial(games_where_not)

# Now, we combine them all
naive_chance = (worst_player_chance) \
	* (not_worst_player_chance) \
	* ways_total_unordered_both

# Print to 3 decimal points
print("{:4.3f}".format(naive_chance))

0.172


0.172 (17.2%) might seem low, but in questions like this, that's actually a really high probability. For comparison, let's look at the chance that he's the worst player in 0, 1, 2, 3, 4, and 5 of the 18 games that he played.

In [11]:
examples_to_consider = 5 + 1 # since we're including 0, add 1
for example in range(examples_to_consider):
	games_where_worst = example
	games_where_not = total_games - games_where_worst
	naive_chance = probability_op_is_worst_player ** games_where_worst \
	* ((1 - probability_op_is_worst_player) ** games_where_not) \
	* math.factorial(total_games) / (math.factorial(games_where_worst) * math.factorial(games_where_not))
	print("Chance of {}: {:4.3f}".format(example, naive_chance))

Chance of 0: 0.018
Chance of 1: 0.081
Chance of 2: 0.172
Chance of 3: 0.230
Chance of 4: 0.215
Chance of 5: 0.151


OP being the worst player in 2 of those 18 games is the third most likely outcome. The most likely outcome, at 23%, is that the OP was the worst player in 3 of his games. 

From this, we can conclude that it's plausible that the OP was the worst player on his team in only 2 games.

## Commenter's Perspective
Since the OP's perspective is plausible, we can tell that the commenter is likely not looking at the question the same way as the OP. Instead, I think that he's asking the question, "What's the expected number of games where OP could have improved?"

If the OP isn't 15k MMR (**M**atch**M**aking **R**ating, which is the result of a modified ELO algorithm to measure skill), then the answer should be 100%, but I don't think that that's what the commenter means. Instead, it seems like he's looking at the OP's average performance, and saying that the OP could have improved when he did worse than his average.

Under that definition, OP is worse than his average in 50% of his games and better than his average in 50% of games. With that prior, what's the chance that the OP could only have improved in 2 of his 18 games?

In [12]:
probability_op_could_have_improved = 0.5
total_games = 18
games_where_could_have_improved = 2
games_where_couldnt = total_games - games_where_could_have_improved
naive_chance = probability_op_could_have_improved ** games_where_could_have_improved \
	* ((1 - probability_op_could_have_improved) ** games_where_couldnt) \
	* math.factorial(total_games) / (math.factorial(games_where_could_have_improved) * math.factorial(games_where_couldnt))
print("{:6.5f}".format(naive_chance))

0.00058


0.00058 (0.058%) is a really low chance; it's about 6 out of 100,000. The commenter is right to be skeptical - the chance that OP only played below his average in 2 games is **really** small. Under the commenter's definition, he's right to be skeptical that the OP was only at fault for two of his losses.

### Synthesis
As a result, neither of them is wrong, and neither of them is right. OP is right because, under his definition, he likely wasn't at fault for a large perecentage of his losses. The commenter is right because, under the his definition, there's a very low chance that OP really only played below average in 2 games.

## Analysis
Now we come to the crux of the issue: whose definition is more **useful**? 

### OP's perspective
OP's definition is more useful for avoiding the negative feelings that come from loss. Out of a hypothetical set of 20 games, on average he'd only feel bad about his performance in 4. When OP goes to examine his performance in those games, every single person in those games will be an example of how he could improve, because he was the worst player on his team. 

However, OP isn't directly learning how to play well - OP's learning how to not play poorly. Because of that, it'll take him a lot more time to learn how to excel in games where he's the best player on his team.

### Commenter's perspective
The commenter's definition is more useful for improving in general, but for players that respond to self-criticism poorly, they run the risk of unbalancing themselves and either getting tilted or getting discouraged. When someone using this definition goes to examine their games, they might find occurrences where they were one of the best players in the game, but because they didn't match the standards that they've set for themselves, they still feel bad.

Under this definition, you aren't playing against other players - you're playing against yourself. If you're the type of person who can handle self-criticism, then this perspective will help you to improve much more quickly. However, it will also require more effort and focus than the alternative perspective.

## Conclusion
The exact definition of "fault" that we use is important, and it can have long-term effects on the way that we learn and improve. Neither of the players in this conversation was **wrong**. Instead, both of them have different perspectives and objectives in playing the game.